pkg/utils/utils.go

```diff
@@ -206,105 +207,187 @@ func DiscoverSriovDevices(withUnsupported bool, storeManager StoreManagerInterfa
-func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState, pfsToConfig map[string]bool) error {
+func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState, pfsToConfig map[string]bool, parallelNicConfig bool) error {
+	if parallelNicConfig {
+		return ConfigSriovInterfacesInParallel(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+	}
 	return ConfigSriovInterfaces(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
 }
 
-func ConfigSriovInterfaces(interfaces []sriovnetworkv1.Interface, ifaceStatuses []sriovnetworkv1.InterfaceExt, pfsToConfig map[string]bool) error {
+func ConfigSriovInterfacesInParallel(interfaces []sriovnetworkv1.Interface, ifaceStatuses []sriovnetworkv1.InterfaceExt, pfsToConfig map[string]bool) error {
+	log.Log.V(2).Info("ConfigSriovInterfacesInParallel(): start sriov configuration")
 	if IsKernelLockdownMode(true) && hasMellanoxInterfacesInSpec(ifaceStatuses, interfaces) {
 		log.Log.Error(nil, "cannot use mellanox devices when in kernel lockdown mode")
 		return fmt.Errorf("cannot use mellanox devices when in kernel lockdown mode")
 	}
-
 	// we are already inside chroot, so we initialize the store as running on host
 	storeManager, err := NewStoreManager(true)
 	if err != nil {
 		return fmt.Errorf("SyncNodeState(): error initializing storeManager: %v", err)
 	}
 
-	for _, ifaceStatus := range ifaceStatuses {
+	// TODO(e0ne): store all errors in SriovNetworkNodeState
+	var result error
+	wg := sync.WaitGroup{}
+	for statusIndex, ifaceStatus := range ifaceStatuses {
 		configured := false
-		for _, iface := range interfaces {
+		for ifaceIndex, iface := range interfaces {
 			if iface.PciAddress == ifaceStatus.PciAddress {
 				configured = true
 
 				if skip := pfsToConfig[iface.PciAddress]; skip {
 					break
 				}
 
-				if !NeedUpdate(&iface, &ifaceStatus) {
-					log.Log.V(2).Info("syncNodeState(): no need update interface", "address", iface.PciAddress)
+				skip, err := checkNeedUpdate(&iface, &ifaceStatus, storeManager)
+				if err != nil {
+					log.Log.Error(err, "ConfigSriovInterfacesInParallel(): failed to check interface")
+					return err
+				}
+				if skip {
+					break
+				}
 
-					// Save the PF status to the host
-					err = storeManager.SaveLastPfAppliedStatus(&iface)
-					if err != nil {
-						log.Log.Error(err, "SyncNodeState(): failed to save PF applied config to host")
-						return err
+				wg.Add(1)
+				go func(iface *sriovnetworkv1.Interface, ifaceStatus *sriovnetworkv1.InterfaceExt) {
+					defer wg.Done()
+					if err := configSriovDevice(iface, ifaceStatus); err != nil {
+						log.Log.Error(err, "ConfigSriovInterfacesInParallel(): fail to configure sriov interface. resetting interface.", "address", iface.PciAddress)
 					}
+				}(&interfaces[ifaceIndex], &ifaceStatuses[statusIndex])
+				// Save the PF status to the host
+				err = storeManager.SaveLastPfAppliedStatus(&iface)
```

Comment:
originally this was saved only if pf configuration suceeded. is that OK to save regardless ?